### PR TITLE
Add preset-create-react-app

### PR DESCRIPTION
**Pull request by mrmckeb**
Resolves https://github.com/storybookjs/storybook/issues/7221

**Additional issues resolved:**
- https://github.com/storybookjs/storybook/issues/4841
- https://github.com/storybookjs/storybook/issues/7220

**In this PR:**
- [x] Warn when used with `preset-typescript` or `preset-scss`.
- [x] Document passing in option `scriptsPackageName`.

**Next steps (before final release):**
- [ ] Add support for custom Babel configs.
- [ ] Add basic unit tests.
- [ ] Create a PR for Storybook: 
  - This preset should be used instead of the existing functionality.
  - On `init`, this preset should be installed if the app matches the criteria.
- [ ] Document migration.
- [ ] Add and document support for `react-docgen-typescript-loader`.
  - Covers https://github.com/storybookjs/presets/pull/21

### CODE REVIEW

File: packages/preset-create-react-app/.storybook/config.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { configure } from '@storybook/react';
+
+const req = require.context('./test-files', true, /.story.(j|t)sx?$/);
```
Comment (shilman) [inline packages/preset-create-react-app/.storybook/config.js]: Let's use `.stories.*` to be consistent?

File: package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "private": true,
   "name": "@storybook/root",
-  "version": "1.0.0",
+  "version": "1.0.0-beta.1",
```
Comment (shilman) [inline package.json]: why?

File: packages/preset-create-react-app/README.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+```js
+module.exports = ['@storybook/preset-create-react-app'];
+```
+
+## Advanced usage
+
+In most cases, this preset will find your `react-scripts` package, even if it's a fork of the offical `react-scripts`.
+
+In the event that it doesn't, you can set the package's name with `scriptsPackageName`.
```
Comment (shilman) [inline packages/preset-create-react-app/README.md]: 💯 this is dope

File: packages/preset-create-react-app/helpers/checkPresets.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
+const { logger } = require('@storybook/node-logger');
+
+const incompatiblePresets = ['@storybook/preset-scss', '@storybook/preset-typescript'];
+
+const checkPresets = () =>
+  incompatiblePresets.forEach(preset => {
+    try {
+      require.resolve(preset);
+      logger.warn(
+        `\`${preset}\` may not be compatible with \`@storybook/preset-create-react-app\``
```
Comment (shilman) [inline packages/preset-create-react-app/helpers/checkPresets.js]: Beautiful 😍

File: packages/preset-create-react-app/helpers/processCraConfig.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        include: [include, storybookConfigDir],
+      };
+      return [...rules, newRule];
+    }
+
+    if (oneOf) {
+      return [
+        ...rules,
+        {
+          oneOf: oneOf.map(oneOfRule => {
```
Comment (shilman) [inline packages/preset-create-react-app/helpers/processCraConfig.js]: What's going on here? Can you add a comment?

File: packages/preset-create-react-app/helpers/reactScriptsPath.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+   * Try to find the `react-scripts` package by name (won't catch forked scripts packages).
+   */
+  try {
+    const scriptsPath = require.resolve('react-scripts');
+    return scriptsPath;
+  } catch (e) {
+    // NOOP
+  }
+
+  return '';
```
Comment (shilman) [inline packages/preset-create-react-app/helpers/reactScriptsPath.js]: null?

File: packages/preset-create-react-app/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+import { Configuration, RuleSetCondition } from 'webpack';
+
+interface Options {
```
Comment (shilman) [inline packages/preset-create-react-app/index.d.ts]: Nice. Do we need these for the other presets too?

File: packages/preset-create-react-app/helpers/checkPresets.js (hunk 4 above)
Comment (shilman) [reply packages/preset-create-react-app/helpers/checkPresets.js]: This checks if the presets are installed. Slightly more accurate if we check if they're used in the user's `presets.js` file? Not a big deal.

File: packages/preset-create-react-app/index.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+checkPresets();
+
+const resolveLoader = {
+  modules: ['node_modules', path.join(reactScriptsPath, 'node_modules')],
+};
+
+// Ensure that loaders are resolved from react-scripts.
+const managerWebpack = (webpackConfig = {}) => ({
+  ...webpackConfig,
+  resolveLoader,
```
Comment (shilman) [inline packages/preset-create-react-app/index.js]: why is this needed for the manager?!

🤔 🤔 🤔 🤔 🤔 🤔 🤔 

This is my only real concern from this PR

File: packages/preset-create-react-app/index.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+const webpack = (webpackConfig = {}, options = {}) => {
+  const storybookConfigDir = path.resolve(options.configDir);
+  let scriptsPath = reactScriptsPath;
+
+  // If the user has provided a package by name, try to resolve it.
+  if (options[SCRIPTS_PACKAGE_OPTION]) {
+    try {
+      scriptsPath = require.resolve(options[SCRIPTS_PACKAGE_OPTION]);
+    } catch (e) {
+      logger.warn(`A \`${SCRIPTS_PACKAGE_OPTION}\` was provided, but couldn't be resolved.`);
```
Comment (shilman) [inline packages/preset-create-react-app/index.js]: Maybe throw here?

File: packages/preset-create-react-app/index.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
+const path = require('path');
+const { logger } = require('@storybook/node-logger');
+const mergePlugins = require('./helpers/mergePlugins');
+const reactScriptsPath = require('./helpers/reactScriptsPath');
```
Comment (shilman) [inline packages/preset-create-react-app/index.js]: should this be a function instead?

File: packages/preset-typescript/README.md [hunk 11]
```diff
@@ -1,6 +1,6 @@
-# Typescript preset for Storybook
+# TypeScript preset for Storybook
```
Comment (shilman) [inline packages/preset-typescript/README.md]: I know this is correct, but it makes my eyes bleed 😭

File: package.json (hunk 2 above)
Comment (mrmckeb) [reply package.json]: Um... this was the wrong file. I meant to do this in the preset. Code review win.

File: packages/preset-create-react-app/.storybook/config.js (hunk 1 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/.storybook/config.js]: Sure, I think I copied this from another preset - but I agree. Will do.

File: packages/preset-create-react-app/helpers/checkPresets.js (hunk 4 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/helpers/checkPresets.js]: Will do.

File: packages/preset-create-react-app/helpers/processCraConfig.js (hunk 5 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/helpers/processCraConfig.js]: It's magic! Will do.

File: packages/preset-create-react-app/index.d.ts (hunk 7 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/index.d.ts]: We seem to have them for some... I think we could move the whole project to TS after we get this out, and then we won't need this.

File: packages/preset-create-react-app/index.js (hunk 8 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/index.js]: Good question. This is the same solution we use now in `@storybook/react` and it allows us to use the `babel-loader` that `react-scripts` ships with - so we avoid conflicts with that version.

File: packages/preset-create-react-app/index.js (hunk 9 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/index.js]: Well, I wasn't sure. It may still work, as we might find the package anyway. So I feel it's safe to just warn?

File: packages/preset-create-react-app/index.js (hunk 10 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/index.js]: I thought about it, but it felt like this solution would be suffice in that you're just importing a path name (if found) - and it's fast in that it only looks once. A function would run multiple times (as we use it multiple times), and then we'd need to memoise.

File: packages/preset-typescript/README.md (hunk 11 above)
Comment (mrmckeb) [reply packages/preset-typescript/README.md]: Hahaha, I'm a little OCD.

File: packages/preset-create-react-app/index.js (hunk 10 above)
Comment (mrmckeb) [reply packages/preset-create-react-app/index.js]: I did it. But it's invoked once only.

**Comment by montogeek** [comment]
Hello, It is possible to use this version of Storybook though an alpha?

**Comment by mrmckeb** [comment]
Hi @montogeek, this is compatible with the latest beta of Storybook - but this preset itself hasn't been published.

@shilman, perhaps we could publish a beta so that people can start testing this preset?

**Comment by shilman** [comment]
@mrmckeb https://www.npmjs.com/package/@storybook/preset-create-react-app

One step ahead of ya...

**Comment by mrmckeb** [comment]
You always are ^^ 😆
